### ISSUE TITLE
h2ph breaks on "function_like_macro (args)"

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#5546]
Migrated from [rt.perl.org#9479](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=9479) (status was 'resolved')

Searchable as RT9479$

**Comment by p5pRT** [comment]
### From prj@po.cwru.edu

This is a bug report for perl from prj@&#8203;multivac\.cwru\.edu\,
generated with the help of perlbug 1\.33 running under perl v5\.6\.1\.




---
h2ph breaks on constructs such as this part of glibc 2\.2\.5's
features\.h&#8203;:
\#if \_\_GNUC\_PREREQ \(2\, 7\) && defined \_\_OPTIMIZE\_\_ \\
&nbsp;    && \!defined \_\_OPTIMIZE\_SIZE\_\_ && \!defined \_\_NO\_INLINE\_\_

It produces this&#8203;:
&nbsp;    if\(\(defined\(&\_\_GNUC\_PREREQ\) ? &\_\_GNUC\_PREREQ : 0\) \(2\, 7\)  && defined \(defined\(&\_\_OPTIMIZE\_\_\) ? &\_\_OPTIMIZE\_\_ : 0\)  && \!defined \(defined\(&\_\_OPTIMIZE\_SIZE\_\_\) ? &\_\_OPTIMIZE\_SIZE\_\_ : 0\)  && \!defined \(defined\(&\_\_NO\_INLINE\_\_\) ? &\_\_NO\_INLINE\_\_ : 0\)\) \{

This patch seems to fix it&#8203;:
<details><summary>Inline Patch</summary>

```diff
--- h2ph.orig   2002-06-06 18:10:00.000000000 -0400
+++ h2ph        2002-06-06 18:09:27.000000000 -0400
@@ -353,8 +353,8 @@
                $new .= '->' if /^[\[\{]/;
            } elsif ($id eq 'defined') {
                $new .= 'defined';
-           } elsif (/^\(/) {
-               s/^\((\w),/("$1",/ if $id =~ /^_IO[WR]*$/i;     # cheat
+           } elsif (/^\s*\(/) {
+               s/^\s*\((\w),/("$1",/ if $id =~ /^_IO[WR]*$/i;  # cheat
                $new .= " &$id";
            } elsif ($isatype{$id}) {
                if ($new =~ /{\s*$/) {

-----------------------------------------------------------------
---

```
</details>
Flags:
&nbsp;    category=utilities
&nbsp;    severity=medium


---
Site configuration information for perl v5\.6\.1&#8203;:

Configured by prj at Thu Jun  6 09&#8203;:38&#8203;:34 EDT 2002\.

Summary of my perl5 \(revision 5\.0 version 6 subversion 1\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=2\.4\.18\, archname=i686\-linux\-64int\-ld
&nbsp;    uname='linux multivac\.cwru\.edu 2\.4\.18 \#1 smp mon may 27 01&#8203;:05&#8203;:55 edt 2002 i686 unknown '
&nbsp;    config\_args='\-Dprefix=/package/host/localhost/perl\-5\.6\.1\-3 \-Uinstallusrbinperl \-Dlocincpth=/package/host/localhost/perl\-5\.6\.1\-3/conf/gdbm/include \-Dloclibpth=/package/host/localhost/perl\-5\.6\.1\-3/conf/gdbm/lib \-Aprepend&#8203;:cppflags= \-Accflags= \-O2  \-Aprepend&#8203;:ldflags= \-O2   \-Wl\,\-R\,/package/host/localhost/perl\-5\.6\.1\-3/conf/gdbm/lib \-Dusemorebits \-de'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
&nbsp;    useperlio=undef d\_sfio=undef uselargefiles=define usesocks=undef
&nbsp;    use64bitint=define use64bitall=undef uselongdouble=define
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-O2 \-fno\-strict\-aliasing \-I/package/host/localhost/perl\-5\.6\.1\-3/conf/gdbm/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;    optimize='\-O2'\,
&nbsp;    cppflags='\-O2 \-fno\-strict\-aliasing \-I/package/host/localhost/perl\-5\.6\.1\-3/conf/gdbm/include'
&nbsp;    ccversion=''\, gccversion='3\.1'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=12345678
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
&nbsp;    ivtype='long long'\, ivsize=8\, nvtype='long double'\, nvsize=12\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=4\, usemymalloc=n\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =' \-O2 \-Wl\,\-R\,/package/host/localhost/perl\-5\.6\.1\-3/conf/gdbm/lib \-L/package/host/localhost/perl\-5\.6\.1\-3/conf/gdbm/lib'
&nbsp;    libpth=/package/host/localhost/perl\-5\.6\.1\-3/conf/gdbm/lib /lib /usr/lib
&nbsp;    libs=\-lnsl \-lgdbm \-ldl \-lm \-lc \-lcrypt \-lutil
&nbsp;    perllibs=\-lnsl \-ldl \-lm \-lc \-lcrypt \-lutil
&nbsp;    libc=/lib/libc\-2\.2\.4\.so\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-rdynamic'
&nbsp;    cccdlflags='\-fpic'\, lddlflags='\-shared \-L/package/host/localhost/perl\-5\.6\.1\-3/conf/gdbm/lib'

Locally applied patches&#8203;:
&nbsp;    



---
@&#8203;INC for perl v5\.6\.1&#8203;:
&nbsp;    /package/host/localhost/perl\-5\.6\.1\-3/lib/5\.6\.1/i686\-linux\-64int\-ld
&nbsp;    /package/host/localhost/perl\-5\.6\.1\-3/lib/5\.6\.1
&nbsp;    /package/host/localhost/perl\-5\.6\.1\-3/lib/site\_perl/5\.6\.1/i686\-linux\-64int\-ld
&nbsp;    /package/host/localhost/perl\-5\.6\.1\-3/lib/site\_perl/5\.6\.1
&nbsp;    /package/host/localhost/perl\-5\.6\.1\-3/lib/site\_perl
&nbsp;    \.



---
Environment for perl v5\.6\.1&#8203;:
&nbsp;    HOME=/home/prj
&nbsp;    LANG \(unset\)
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=/home/prj/command&#8203;:/command
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    SHELL=/command/bashrc

**Comment by p5pRT** [comment]
### From @jhi

On Thu\, Jun 06\, 2002 at 06&#8203;:12&#8203;:28PM \-0400\, Paul Jarc wrote&#8203;:
> This is a bug report for perl from prj@&#8203;multivac\.cwru\.edu\,
> generated with the help of perlbug 1\.33 running under perl v5\.6\.1\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> h2ph breaks on constructs such as this part of glibc 2\.2\.5's
> features\.h&#8203;:
> \#if \_\_GNUC\_PREREQ \(2\, 7\) && defined \_\_OPTIMIZE\_\_ \\
>     && \!defined \_\_OPTIMIZE\_SIZE\_\_ && \!defined \_\_NO\_INLINE\_\_
> 
> It produces this&#8203;:
>     if\(\(defined\(&\_\_GNUC\_PREREQ\) ? &\_\_GNUC\_PREREQ : 0\) \(2\, 7\)  && defined \(defined\(&\_\_OPTIMIZE\_\_\) ? &\_\_OPTIMIZE\_\_ : 0\)  && \!defined \(defined\(&\_\_OPTIMIZE\_SIZE\_\_\) ? &\_\_OPTIMIZE\_SIZE\_\_ : 0\)  && \!defined \(defined\(&\_\_NO\_INLINE\_\_\) ? &\_\_NO\_INLINE\_\_ : 0\)\) \{
> 
> This patch seems to fix it&#8203;:
> \-\-\- h2ph\.orig   2002\-06\-06 18&#8203;:10&#8203;:00\.000000000 \-0400
> \+\+\+ h2ph        2002\-06\-06 18&#8203;:09&#8203;:27\.000000000 \-0400
> @&#8203;@&#8203; \-353\,8 \+353\,8 @&#8203;@&#8203;
>                 $new \.= '\->' if /^\[\\\[\\\{\]/;
>             \} elsif \($id eq 'defined'\) \{
>                 $new \.= 'defined';
> \-           \} elsif \(/^\\\(/\) \{
> \-               s/^\\\(\(\\w\)\,/\("$1"\,/ if $id =~ /^\_IO\[WR\]\*$/i;     \# cheat
> \+           \} elsif \(/^\\s\*\\\(/\) \{
> \+               s/^\\s\*\\\(\(\\w\)\,/\("$1"\,/ if $id =~ /^\_IO\[WR\]\*$/i;  \# cheat

I don't think that's quite right since that results in

&nbsp;    if\( &\_\_GNUC\_PREREQ \(2\, 7\)  && defined \(defined\(&\_\_OPTIMIZE\_\_\) ? &\_\_OPTIMIZE\_\_ : 0\)  && \!defined \(defined\(&\_\_OPTIMIZE\_SIZE\_\_\) ? &\_\_OPTIMIZE\_SIZE\_\_ : 0\)  && \!defined \(defined\(&\_\_NO\_INLINE\_\_\) ? &\_\_NO\_INLINE\_\_ : 0\)\) \{
&nbsp;    \}

In other words\, there's no defined\(&\_\_GNU\_PREREQ\) anywhere\. 

>                 $new \.= " &$id";
>             \} elsif \($isatype\{$id\}\) \{
>                 if \($new =~ /\{\\s\*$/\) \{
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

\-\- 
$jhi\+\+; \# http&#8203;://www\.iki\.fi/jhi/
&nbsp;        \# There is this special biologist word we use for 'stable'\.
&nbsp;        \# It is 'dead'\. \-\- Jack Cohen

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Jarkko Hietaniemi \<jhi@&#8203;iki\.fi> wrote&#8203;:
> On Thu\, Jun 06\, 2002 at 06&#8203;:12&#8203;:28PM \-0400\, Paul Jarc wrote&#8203;:
>> \-           \} elsif \(/^\\\(/\) \{
>> \-               s/^\\\(\(\\w\)\,/\("$1"\,/ if $id =~ /^\_IO\[WR\]\*$/i;     \# cheat
>> \+           \} elsif \(/^\\s\*\\\(/\) \{
>> \+               s/^\\s\*\\\(\(\\w\)\,/\("$1"\,/ if $id =~ /^\_IO\[WR\]\*$/i;  \# cheat
>
> I don't think that's quite right since that results in
>
>     if\( &\_\_GNUC\_PREREQ \(2\, 7\)  && defined \(defined\(&\_\_OPTIMIZE\_\_\) ? &\_\_OPTIMIZE\_\_ : 0\)  && \!defined \(defined\(&\_\_OPTIMIZE\_SIZE\_\_\) ? &\_\_OPTIMIZE\_SIZE\_\_ : 0\)  && \!defined \(defined\(&\_\_NO\_INLINE\_\_\) ? &\_\_NO\_INLINE\_\_ : 0\)\) \{
>     \}

If that is wrong\, then h2ph is broken even without the patch\.  The
only effect of the patch is to make h2ph treat "macro \(args\)" the same
way it already treats "macro\(args\)"\.  Those constructs should indeed
be treated the same way\.  Exactly how they should be treated\, I'll
leave to someone else\.

> In other words\, there's no defined\(&\_\_GNU\_PREREQ\) anywhere\.

Nor should there be\, AFAICT; that would call the \_\_GNU\_PREREQ function
with no arguments\, while it is expecting two\.


paul

**Comment by p5pRT** [comment]
### From @jhi

\*cough\*  In any case\, it seems that somebody has already done the
exactly same h2ph patching in the past for Perl 5\.8\.0\-to\-be\.\.\.

\-\- 
$jhi\+\+; \# http&#8203;://www\.iki\.fi/jhi/
&nbsp;        \# There is this special biologist word we use for 'stable'\.
&nbsp;        \# It is 'dead'\. \-\- Jack Cohen

**Comment by p5pRT** [comment]
@jhi - Status changed from 'open' to 'resolved'
